### PR TITLE
chore: Implement standard rbac.Subject to be reused everywhere 

### PR DESCRIPTION
**Pull request by Emyrk**
**This is purely a type refactor, and does not change any behavior**

# What this does

Before this PR, the rbac package expected the caller to hold the various `Subject` fields, and pass them in as individual arguments: `(subjectID string, roleNames ExpandableRoles, scope ScopeName, groups []string)`. This was ok in the beginning when it was just `(ID, Roles)`, but it has grown to `scope` and `groups` as well.

This change ensures all subject fields are tightly coupled, and less likely to be mistakenly used. It also simplifies using `rbac` since the arguments are now `Autorize(ctx, subject, action, object)`. 

The only downside is that logging still requires poking into the subject struct. But usage of the rbac package now takes less arguments.

# What this enables

For `authquerier` interface we wanted to enable the use of a `System` actor. This actor needs to have custom roles that should not be available to the user. This new `Subject` struct accepts both user defined roles, `RoleNames` and explicit `Roles`.

This allows us to create custom roles that will never be returned via an api or any other exporting of roles (docs, enums, etc).

Minor bonus, this allows unit tests to call the `Authorize` directly as it was using explicit Roles.

### CODE REVIEW

File: coderd/rbac/authz.go [hunk 1]
```diff
@@ -17,24 +17,34 @@ import (
...
-	// names of the expanded roles.
-	Names() []string
+// Subject is a struct that contains all the elements of a subject in an rbac
+// authorize.
+type Subject struct {
+	ID     string
+	Roles  ExpandableRoles
+	Groups []string
+	Scope  ExpandableScope
+}
```
Comment (Emyrk) [inline coderd/rbac/authz.go:27]: This is the added type shared by all callers now.

File: coderd/rbac/authz.go [hunk 2]
```diff
@@ -17,24 +17,34 @@ import (
...
+		return []string{}
+	}
+	return s.Roles.Names()
 }
 
 type Authorizer interface {
-	ByRoleName(ctx context.Context, subjectID string, roleNames ExpandableRoles, scope ScopeName, groups []string, action Action, object Object) error
-	PrepareByRoleName(ctx context.Context, subjectID string, roleNames ExpandableRoles, scope ScopeName, groups []string, action Action, objectType string) (PreparedAuthorized, error)
+	Authorize(ctx context.Context, subject Subject, action Action, object Object) error
+	Prepare(ctx context.Context, subject Subject, action Action, objectType string) (PreparedAuthorized, error)
```
Comment (Emyrk) [inline coderd/rbac/authz.go:47]: This is the interface of the package changing to use less arguments.

File: coderd/rbac/authz.go [hunk 3]
```diff
@@ -207,18 +218,9 @@ func (a RegoAuthorizer) ByRoleName(ctx context.Context, subjectID string, roleNa
...
-	if err != nil {
-		return err
-	}
-
-	scopeRole, err := ExpandScope(scope)
-	if err != nil {
-		return err
-	}
+	err := a.authorize(ctx, subject, action, object)
+	span.SetAttributes(attribute.Bool("authorized", err == nil))
```
Comment (Emyrk) [inline coderd/rbac/authz.go:222]: This is better than an event.

File: coderd/authorize.go [hunk 4]
```diff
@@ -217,7 +219,7 @@ func (api *API) checkAuthorization(rw http.ResponseWriter, r *http.Request) {
 			obj = dbObj.RBACObject()
 		}
 
-		err := api.Authorizer.ByRoleName(ctx, auth.ID.String(), auth.Roles, auth.Scope.ToRBAC(), auth.Groups, rbac.Action(v.Action), obj)
+		err := api.Authorizer.Authorize(ctx, auth.Actor, rbac.Action(v.Action), obj)
```
Comment (johnstcn) [inline coderd/authorize.go:222]: This is so much nicer and more readable ❤️

File: coderd/rbac/authz.go [hunk 5]
```diff
@@ -17,24 +17,34 @@ import (
...
+type Subject struct {
+	ID     string
+	Roles  ExpandableRoles
+	Groups []string
+	Scope  ExpandableScope
+}
+
+// SafeScopeName prevent nil pointer dereference.
+func (s Subject) SafeScopeName() string {
+	if s.Scope == nil {
```
Comment (johnstcn) [inline coderd/rbac/authz.go:31]: Looking at RegoAuthorizer.Authorize, wouldn't this be a developer error?

**Review by johnstcn — APPROVED** [review]
👍

File: coderd/rbac/authz.go (hunk 5 above)
Comment (Emyrk) [reply coderd/rbac/authz.go:31]: It would be a developer error.

**Pull request merged**